src/utils.c

```diff
@@ -363,3 +363,84 @@ SEXP coerceUtf8IfNeeded(SEXP x) {
   return(ans);
 }
 
+// if (length(x)) length(x[[1L]]) else 0L
+int NROW(SEXP x) {
+  if (!LENGTH(x))
+    return 0; // # nocov # not yet reached from anywhere, cbindlist uses it but escapes for !NCOL(x)
+  return length(VECTOR_ELT(x, 0));
+}
+
+// length(x)
+int NCOL(SEXP x) {
+  return LENGTH(x);
+}
+
+/*// c("data.table","data.frame")
+static SEXP char2_dtdf() {
+  SEXP char2_dtdf = PROTECT(allocVector(STRSXP, 2));
+  SET_STRING_ELT(char2_dtdf, 0, char_datatable);
+  SET_STRING_ELT(char2_dtdf, 1, char_dataframe);
+  UNPROTECT(1);
+  return char2_dtdf;
+}
+
+// .set_row_names(x)
+static SEXP set_row_names(int n) {
+  SEXP ans = R_NilValue;
+  if (n) {
+    ans = PROTECT(allocVector(INTSXP, 2));
+    INTEGER(ans)[0] = NA_INTEGER;
+    INTEGER(ans)[1] = -n;
+  } else {
+    ans = PROTECT(allocVector(INTSXP, 0));
+  }
+  UNPROTECT(1);
+  return ans;
+}
+
+// setDT(x) ## not in-place!
+SEXP setDT(SEXP x) {
+  if (!isNewList(x))
+    error("internal error: C setDT should be called only on a list"); // # nocov
+  setAttrib(x, R_ClassSymbol, char2_dtdf());
+  setAttrib(x, sym_rownames, set_row_names(NROW(x)));
+  return alloccolwrapper(x, GetOption(sym_alloccol, R_NilValue), GetOption(sym_verbose, R_NilValue));
+}*/
+
+// inherits(x, "data.table")
+bool isDataTable(SEXP x) {
+  return INHERITS(x, char_datatable);
+}
+
+// inherits(x, "data.frame")
+bool isDataFrame(SEXP x) {
+  return INHERITS(x, char_dataframe);
+}
+
+// if (length(x)>1L) length(unique(vapply(x, length, 0L)))==1L else TRUE
+static inline bool equalLens(SEXP x) {
+  int n = LENGTH(x);
+  if (n < 2)
+    return true;
+  for (int i=1; i<n; ++i) {
+    if (xlength(VECTOR_ELT(x, i)) != xlength(VECTOR_ELT(x, i-1)))
```

Comment:
this runs `xlength(VECTOR_ELT(.))` on all but the first/last elements twice right? it's all pointers but still